File: app/src/main/java/com/ty/zbpet/ui/activity/material/ArrivalInDoneDetailActivity.java
package com.ty.zbpet.ui.activity.material;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.InputType;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.bigkoo.pickerview.listener.OnTimeSelectListener;
import com.ty.zbpet.R;
import com.ty.zbpet.bean.ResponseInfo;
import com.ty.zbpet.bean.material.MaterialDetailsOut;
import com.ty.zbpet.bean.material.MaterialDoneSave;
import com.ty.zbpet.net.HttpMethods;
import com.ty.zbpet.presenter.material.MaterialPresenter;
import com.ty.zbpet.presenter.material.MaterialUiListInterface;
import com.ty.zbpet.ui.adapter.SwipeItemLayout;
import com.ty.zbpet.ui.adapter.material.MaterialDoneDetailAdapter;
import com.ty.zbpet.ui.base.BaseActivity;
import com.ty.zbpet.ui.widght.SpaceItemDecoration;
import com.ty.zbpet.util.CodeConstant;
import com.ty.zbpet.util.DataUtils;
import com.ty.zbpet.util.ResourceUtil;
import com.ty.zbpet.util.ZBUiUtils;
import com.zhouyou.http.exception.ApiException;
import com.zhouyou.http.subsciber.BaseSubscriber;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;
import okhttp3.RequestBody;

/**
 * @author TY on 2018/11/14.
 * <p>
 * 已办 详情
 */
public class ArrivalInDoneDetailActivity extends BaseActivity implements MaterialUiListInterface<MaterialDetailsOut.ListBean> {

    @BindView(R.id.rv_in_storage_detail)
    RecyclerView detailRc;
    @BindView(R.id.tv_time)
    TextView tvTime;

    @BindView(R.id.et_desc)
    EditText etDesc;

    /**
     * 时间选择
     */
    private String selectTime;


    private String orderId;
    private String mInWarehouseOrderId;
    private String warehouseId;
    private String sapOrderNo;
    private String positionId;

    private MaterialDoneDetailAdapter adapter;
    private MaterialPresenter materialPresenter = new MaterialPresenter(this);


    @Override
    protected void onBaseCreate(Bundle savedInstanceState) {
        SimpleDateFormat sdf = new SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA);
        selectTime = sdf.format(new Date());
        tvTime.setText(selectTime);
    }

    @Override
    protected int getActivityLayout() {
        return R.layout.activity_content_row_two;
    }

    @Override
    protected void initOneData() {

        mInWarehouseOrderId = getIntent().getStringExtra("mInWarehouseOrderId");
        sapOrderNo = getIntent().getStringExtra("sapOrderNo");
        warehouseId = getIntent().getStringExtra("warehouseId");
        orderId = getIntent().getStringExtra("orderId");
    }

    @Override
    protected void initTwoView() {

        findViewById(R.id.add_ship).setVisibility(View.GONE);
        TextView titleName = findViewById(R.id.in_storage_detail);
        titleName.setText("到货明细");
        etDesc.setInputType(InputType.TYPE_NULL);

        initToolBar(R.string.material_reversal, new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                materialDoneInSave(initRequestBody());

            }
        });


    }

    /**
     * 已办 保存
     * @param body
     */
    public void materialDoneInSave(RequestBody body) {
        HttpMethods.getInstance().materialDoneInSave(new SingleObserver<ResponseInfo>() {
            @Override
            public void onError(Throwable e) {
                ZBUiUtils.showToast(e.getMessage());
            }

            @Override
            public void onSubscribe(Disposable d) {

            }
            @Override
            public void onSuccess(ResponseInfo responseInfo) {

                if (CodeConstant.SERVICE_SUCCESS.equals(responseInfo.getTag())) {
                    ZBUiUtils.showToast(responseInfo.getMessage());
                    finish();
                } else {
                    ZBUiUtils.showToast(responseInfo.getMessage());
                }
            }

        }, body);

    }

    /**
     * 初始化 请求参数
     *
     * @return
     */
    private RequestBody initRequestBody() {

        MaterialDoneSave bean = new MaterialDoneSave();

        bean.setWarehouseId(warehouseId);
        bean.setOrderId(mInWarehouseOrderId);
        bean.setSapOrderNo(sapOrderNo);
        bean.setPositionId(positionId);
        bean.setOrderId(orderId);

        String json = DataUtils.toJson(bean, 1);

        return RequestBody.create(okhttp3.MediaType.parse("application/json;charset=UTF-8"), json);

    }

    @Override
    protected void onStart() {
        super.onStart();
        materialPresenter.fetchDoneMaterialDetails(orderId);
    }

    @Override
    public void showMaterial(List<MaterialDetailsOut.ListBean> list) {

        positionId = list.get(0).getPositionId();

        if (adapter == null) {
            LinearLayoutManager manager = new LinearLayoutManager(ResourceUtil.getContext());
            detailRc.addItemDecoration(new SpaceItemDecoration(ResourceUtil.dip2px(10), false));
            detailRc.setLayoutManager(manager);

            // TODO 侧滑删除
            // detailRc.addOnItemTouchListener(new SwipeItemLayout.OnSwipeItemTouchListener(this));
            adapter = new MaterialDoneDetailAdapter(this, R.layout.item_material_done_detail, list);
            detailRc.setAdapter(adapter);
            adapter.setOnItemClickListener(new MaterialDoneDetailAdapter.OnItemClickListener() {
                @Override
                public void onItemClick(View view, RecyclerView.ViewHolder holder, int position) {
                    View rlDetail = holder.itemView.findViewById(R.id.view_gone);
                    ImageView ivArrow = holder.itemView.findViewById(R.id.iv_arrow);

                    if (rlDetail.getVisibility() == View.VISIBLE) {
                        rlDetail.setVisibility(View.GONE);
                        ivArrow.setImageResource(R.mipmap.ic_collapse);
                    } else {
                        rlDetail.setVisibility(View.VISIBLE);
                        ivArrow.setImageResource(R.mipmap.ic_expand);
                    }
                }

                @Override
                public boolean onItemLongClick(View view, RecyclerView.ViewHolder holder, int position) {
                    return false;
                }
            });
        }

    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }
}

File: app/src/main/java/com/ty/zbpet/ui/activity/material/ArrivalInDoneDetailActivity.kt
package com.ty.zbpet.ui.activity.material

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.widget.CheckBox
import com.ty.zbpet.R
import com.ty.zbpet.bean.CarPositionNoData
import com.ty.zbpet.bean.material.MaterialDetails
import com.ty.zbpet.bean.material.MaterialDoneSave
import com.ty.zbpet.constant.CodeConstant
import com.ty.zbpet.net.RequestBodyJson
import com.ty.zbpet.presenter.material.MaterialPresenter
import com.ty.zbpet.presenter.material.MaterialUiListInterface
import com.ty.zbpet.ui.adapter.material.MaterialDoneDetailAdapter
import com.ty.zbpet.base.BaseActivity
import com.ty.zbpet.ui.widght.ShowDialog
import com.ty.zbpet.ui.widght.SpaceItemDecoration
import com.ty.zbpet.util.DataUtils
import com.ty.zbpet.util.ResourceUtil
import com.ty.zbpet.util.ZBUiUtils
import com.xiasuhuei321.loadingdialog.view.LoadingDialog
import kotlinx.android.synthetic.main.activity_content_reversal.*
import okhttp3.RequestBody
import java.util.*

/**
 * @author TY on 2018/11/14.
 *
 *
 * 已办 详情
 */
class ArrivalInDoneDetailActivity : BaseActivity(), MaterialUiListInterface<MaterialDetails.ListBean> {
    override val activityLayout: Int
        get() = R.layout.activity_content_reversal

    private lateinit var orderId: String
    private lateinit var sapOrderNo: String
    private lateinit var supplierName: String

    private lateinit var listBean: List<MaterialDetails.ListBean>

    private lateinit var adapter: MaterialDoneDetailAdapter
    private val materialPresenter = MaterialPresenter(this)


    override fun onBaseCreate(savedInstanceState: Bundle?) {
    }

    override fun initOneData() {

        sapOrderNo = intent.getStringExtra("sapOrderNo")
        supplierName = intent.getStringExtra("supplierName")
        orderId = intent.getStringExtra("orderId")
    }

    override fun initTwoView() {

        initToolBar(R.string.purchase_house_reversal)
        bt_reversal.setOnClickListener { materialDoneInSave(initRequestBody()) }
    }

    /**
     * 已办 保存
     * @param body
     */
    private fun materialDoneInSave(body: RequestBody?) {

        if (body == null) {
            return
        }
        materialPresenter.materialDoneInSave(body)

    }

    /**
     * 初始化 请求参数
     *
     * @return
     */
    private fun initRequestBody(): RequestBody? {

        val data = MaterialDoneSave()
        val list = ArrayList<MaterialDoneSave.ListBean>()
        val size = listBean.size

        for (i in 0 until size ){
            val view = recycler_reversal.getChildAt(i)
            val checkBox = view.findViewById<CheckBox>(R.id.check)
            if (checkBox.isChecked) {
                val bean = MaterialDoneSave.ListBean()
                bean.id = listBean[i].id
                bean.orderId = orderId
                list.add(bean)
            }
        }

        if(list.size == 0){
            ZBUiUtils.showWarning("请选择您要冲销的列表")
            return null
        }

        data.list = list
        data.supplierName = supplierName
        data.moveType = "106"

        val json = DataUtils.toJson(data, 1)
        return RequestBodyJson.requestBody(json)

    }

    override fun onStart() {
        super.onStart()
        materialPresenter.fetchDoneMaterialDetails(orderId)
    }

    override fun showMaterial(list: List<MaterialDetails.ListBean>) {
        listBean = list

        val manager = LinearLayoutManager(ResourceUtil.getContext())
        recycler_reversal.addItemDecoration(SpaceItemDecoration(ResourceUtil.dip2px(CodeConstant.ITEM_DECORATION), false))
        recycler_reversal.layoutManager = manager

        // TODO 侧滑删除
        // detailRc.addOnItemTouchListener(new SwipeItemLayout.OnSwipeItemTouchListener(this));
        adapter = MaterialDoneDetailAdapter(this, R.layout.item_reversal_check, list)
        recycler_reversal.adapter = adapter

    }
    override fun showCarSuccess(position: Int, carData: CarPositionNoData?) {
    }

    override fun saveSuccess() {
        finish()
    }

    override fun showError(msg: String?) {
        ZBUiUtils.showError(msg)
    }

    private var dialog: LoadingDialog? = null
    override fun showLoading() {
        dialog = ShowDialog.showFullDialog(this@ArrivalInDoneDetailActivity, "保存中")
    }

    override fun hideLoading() {
        dialog?.close()
    }
}
